<a href="https://colab.research.google.com/github/dhanajustbegining/AI-Learning/blob/main/week_1_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
print("Hello Vrddhi")

Hello Vrddhi


Alright — let’s go step-by-step and calculate each metric carefully.

We’re given:

* **TP (True Positives)** = 80
* **TN (True Negatives)** = 50
* **FP (False Positives)** = 20
* **FN (False Negatives)** = 10

---

## **1. Precision**

Formula:

$$
\text{Precision} = \frac{TP}{TP + FP}
$$

Substitute:

$$
\text{Precision} = \frac{80}{80 + 20} = \frac{80}{100} = 0.8
$$

✅ **Precision = 0.8 (80%)**

---

## **2. Recall**

Formula:

$$
\text{Recall} = \frac{TP}{TP + FN}
$$

Substitute:

$$
\text{Recall} = \frac{80}{80 + 10} = \frac{80}{90} \approx 0.8889
$$

✅ **Recall ≈ 0.8889 (88.89%)**

---

## **3. F1-score**

Formula:

$$
\text{F1} = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}
$$

Substitute:

$$
\text{F1} = 2 \times \frac{0.8 \times 0.8889}{0.8 + 0.8889}
$$

$$
\text{F1} = 2 \times \frac{0.71112}{1.6889} \approx 2 \times 0.4211 \approx 0.842
$$

✅ **F1-score ≈ 0.842 (84.2%)**

---

## **4. Accuracy**

Formula:

$$
\text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}
$$

Substitute:

$$
\text{Accuracy} = \frac{80 + 50}{80 + 50 + 20 + 10} = \frac{130}{160} = 0.8125
$$

✅ **Accuracy = 0.8125 (81.25%)**

---

### **Final Answers**

* **Precision:** 0.8 (80%)
* **Recall:** ≈ 0.8889 (88.89%)
* **F1-score:** ≈ 0.842 (84.2%)
* **Accuracy:** 0.8125 (81.25%)

---

If you’d like, I can also give you a **Python snippet** so you can calculate these for any TP, TN, FP, FN values without doing them manually.
Do you want me to prepare that?


In [ ]:
# Function to calculate metrics
def calculate_metrics(TP, TN, FP, FN):
    # Precision
    precision = TP / (TP + FP)

    # Recall
    recall = TP / (TP + FN)

    # F1-score
    f1_score = 2 * (precision * recall) / (precision + recall)

    # Accuracy
    accuracy = (TP + TN) / (TP + TN + FP + FN)

    return precision, recall, f1_score, accuracy


# Given values
TP = 80
TN = 50
FP = 20
FN = 10

# Calculate metrics
precision, recall, f1, accuracy = calculate_metrics(TP, TN, FP, FN)

# Display results
print(f"Precision: {precision:.4f} ({precision*100:.2f}%)")
print(f"Recall:    {recall:.4f} ({recall*100:.2f}%)")
print(f"F1-score:  {f1:.4f} ({f1*100:.2f}%)")
print(f"Accuracy:  {accuracy:.4f} ({accuracy*100:.2f}%)")


Precision: 0.8000 (80.00%)
Recall:    0.8889 (88.89%)
F1-score:  0.8421 (84.21%)
Accuracy:  0.8125 (81.25%)


Alright — let’s calculate for **both models** step-by-step so we don’t make mistakes.

---

## **Given Data**

**Model A**

* TP = 120
* TN = 150
* FP = 30
* FN = 20

**Model B**

* TP = 140
* TN = 130
* FP = 40
* FN = 10

---

## **Formulas**

* **Precision** = TP / (TP + FP)
* **Recall** = TP / (TP + FN)
* **F1-score** = 2 × (Precision × Recall) / (Precision + Recall)

---

### **1. Model A**

**Precision:**

$$
\text{Precision}_A = \frac{120}{120 + 30} = \frac{120}{150} = 0.8
$$

**Recall:**

$$
\text{Recall}_A = \frac{120}{120 + 20} = \frac{120}{140} \approx 0.8571
$$

**F1-score:**

$$
\text{F1}_A = 2 \times \frac{0.8 \times 0.8571}{0.8 + 0.8571}
= 2 \times \frac{0.68568}{1.6571} \approx 0.8276
$$

✅ **Model A**

* Precision = 0.8000 (80%)
* Recall ≈ 0.8571 (85.71%)
* F1-score ≈ 0.8276 (82.76%)

---

### **2. Model B**

**Precision:**

$$
\text{Precision}_B = \frac{140}{140 + 40} = \frac{140}{180} \approx 0.7778
$$

**Recall:**

$$
\text{Recall}_B = \frac{140}{140 + 10} = \frac{140}{150} \approx 0.9333
$$

**F1-score:**

$$
\text{F1}_B = 2 \times \frac{0.7778 \times 0.9333}{0.7778 + 0.9333}
= 2 \times \frac{0.72593}{1.7111} \approx 0.8485
$$

✅ **Model B**

* Precision ≈ 0.7778 (77.78%)
* Recall ≈ 0.9333 (93.33%)
* F1-score ≈ 0.8485 (84.85%)

---

## **3. Comparison**

| Metric    | Model A | Model B | Better Model |
| --------- | ------- | ------- | ------------ |
| Precision | 0.8000  | 0.7778  | **A**        |
| Recall    | 0.8571  | 0.9333  | **B**        |
| F1-score  | 0.8276  | 0.8485  | **B**        |

---

✅ **Summary:**

* **Precision:** Model A is better (fewer false positives).
* **Recall:** Model B is better (fewer false negatives).
* **F1-score:** Model B is slightly better overall.

---

If you like, I can give you a **Python function** to compare *any* two models’ precision, recall, and F1-score instantly. That way you won’t need to calculate manually again.
Do you want me to make that?


In [ ]:
def calculate_metrics(TP, TN, FP, FN):
    """Calculate precision, recall, and F1-score."""
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)
    return precision, recall, f1_score

def compare_models(modelA, modelB):
    # Unpack model data
    TP_A, TN_A, FP_A, FN_A = modelA
    TP_B, TN_B, FP_B, FN_B = modelB

    # Calculate metrics
    precision_A, recall_A, f1_A = calculate_metrics(TP_A, TN_A, FP_A, FN_A)
    precision_B, recall_B, f1_B = calculate_metrics(TP_B, TN_B, FP_B, FN_B)

    # Display results
    print("Model A:")
    print(f" Precision: {precision_A:.4f} ({precision_A*100:.2f}%)")
    print(f" Recall:    {recall_A:.4f} ({recall_A*100:.2f}%)")
    print(f" F1-score:  {f1_A:.4f} ({f1_A*100:.2f}%)\n")

    print("Model B:")
    print(f" Precision: {precision_B:.4f} ({precision_B*100:.2f}%)")
    print(f" Recall:    {recall_B:.4f} ({recall_B*100:.2f}%)")
    print(f" F1-score:  {f1_B:.4f} ({f1_B*100:.2f}%)\n")

    # Comparison
    print("Better Model by Metric:")
    print(f" Precision: {'Model A' if precision_A > precision_B else 'Model B'}")
    print(f" Recall:    {'Model A' if recall_A > recall_B else 'Model B'}")
    print(f" F1-score:  {'Model A' if f1_A > f1_B else 'Model B'}")


# Example usage with your data
modelA_data = (120, 150, 30, 20)  # TP, TN, FP, FN
modelB_data = (140, 130, 40, 10)

compare_models(modelA_data, modelB_data)


Model A:
 Precision: 0.8000 (80.00%)
 Recall:    0.8571 (85.71%)
 F1-score:  0.8276 (82.76%)

Model B:
 Precision: 0.7778 (77.78%)
 Recall:    0.9333 (93.33%)
 F1-score:  0.8485 (84.85%)

Better Model by Metric:
 Precision: Model A
 Recall:    Model B
 F1-score:  Model B


Alright, let’s work through this step-by-step so we get everything correct.

---

## **Given Data**

* Total images = 200
* TP = 70
* TN = 100
* FP = 10
* FN = 20

---

### **1. Accuracy**

**Formula:**

$$
\text{Accuracy} = \frac{TP + TN}{\text{Total samples}}
$$

**Calculation:**

$$
\text{Accuracy} = \frac{70 + 100}{200} = \frac{170}{200} = 0.85
$$

✅ **Accuracy = 0.85 (85%)**

---

### **2. Precision vs. Recall**

* **Precision** = TP / (TP + FP)

$$
\text{Precision} = \frac{70}{70 + 10} = \frac{70}{80} = 0.875 \ (87.5\%)
$$

* **Recall** = TP / (TP + FN)

$$
\text{Recall} = \frac{70}{70 + 20} = \frac{70}{90} \approx 0.7778 \ (77.78\%)
$$

---

### **Interpretation**

* **Precision (87.5%)** is higher than **Recall (77.78%)**.
* Higher precision means **fewer false positives** — the car rarely mistakes a non-pedestrian for a pedestrian.
* Lower recall means **more false negatives** — the car misses some actual pedestrians.

---

**Conclusion:**
The model is **better at avoiding false positives** than avoiding false negatives.
However, in a pedestrian detection scenario, missing pedestrians (false negatives) could be more dangerous than a few extra false positives.

---

If you want, I can show you **how to visualize this with a confusion matrix** so the trade-off between false positives and false negatives is easier to see. Would you like me to prepare that?


In [ ]:
def evaluate_model(TP, TN, FP, FN, total_samples):
    # Calculate metrics
    accuracy = (TP + TN) / total_samples
    precision = TP / (TP + FP) if (TP + FP) > 0 else 0
    recall = TP / (TP + FN) if (TP + FN) > 0 else 0

    # Display results
    print(f"Accuracy:  {accuracy:.4f} ({accuracy*100:.2f}%)")
    print(f"Precision: {precision:.4f} ({precision*100:.2f}%)")
    print(f"Recall:    {recall:.4f} ({recall*100:.2f}%)")

    # Interpretation
    if precision > recall:
        print("\nModel is better at avoiding False Positives.")
    elif recall > precision:
        print("\nModel is better at avoiding False Negatives.")
    else:
        print("\nModel performs equally for avoiding False Positives and False Negatives.")

# Example usage with your data
evaluate_model(TP=70, TN=100, FP=10, FN=20, total_samples=200)


Accuracy:  0.8500 (85.00%)
Precision: 0.8750 (87.50%)
Recall:    0.7778 (77.78%)

Model is better at avoiding False Positives.


Let’s go step-by-step so the math is clear.

---

## **Given Data**

* Total samples = 1000
* TP = 400
* TN = 450
* FP = 50
* FN = 100

---

### **1. Metric Calculations**

#### **Precision**

$$
\text{Precision} = \frac{TP}{TP + FP} = \frac{400}{400 + 50} = \frac{400}{450} \approx 0.8889 \ (88.89\%)
$$

#### **Recall**

$$
\text{Recall} = \frac{TP}{TP + FN} = \frac{400}{400 + 100} = \frac{400}{500} = 0.8 \ (80\%)
$$

#### **F1-score**

$$
\text{F1} = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}
$$

$$
\text{F1} = 2 \times \frac{0.8889 \times 0.8}{0.8889 + 0.8} = 2 \times \frac{0.7111}{1.6889} \approx 0.8421 \ (84.21\%)
$$

#### **Accuracy**

$$
\text{Accuracy} = \frac{TP + TN}{\text{Total Samples}} = \frac{400 + 450}{1000} = \frac{850}{1000} = 0.85 \ (85\%)
$$

---

✅ **Summary Table**

| Metric    | Value  |
| --------- | ------ |
| Precision | 88.89% |
| Recall    | 80.00% |
| F1-score  | 84.21% |
| Accuracy  | 85.00% |

---

### **2. Imbalanced Dataset Impact**

If the dataset were **imbalanced** — for example, **80% negative** and **20% positive**:

* Accuracy could still be **high** even if the model performs poorly for the minority (positive) class.
* Example: If the model simply predicts **"negative" for everything**:

  * Accuracy = 80% (because it gets all negatives correct)
  * But **Precision, Recall, and F1-score** for the positive class would be **0**, showing no real predictive power for positives.

**Conclusion:**

* **Accuracy is not reliable** in imbalanced datasets.
* Metrics like **Precision, Recall, F1-score, ROC-AUC** are better for evaluating performance in such cases.

---

If you want, I can **write a Python function** that calculates all these metrics and also simulates the effect of imbalance on accuracy.
Do you want me to prepare that?


In [ ]:
def evaluate_binary_classifier(TP, TN, FP, FN, total_samples, imbalance_ratio=None):
    # Metrics
    precision = TP / (TP + FP) if (TP + FP) > 0 else 0
    recall = TP / (TP + FN) if (TP + FN) > 0 else 0
    f1_score = (2 * precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
    accuracy = (TP + TN) / total_samples

    print("=== Metrics ===")
    print(f"Precision: {precision:.4f} ({precision*100:.2f}%)")
    print(f"Recall:    {recall:.4f} ({recall*100:.2f}%)")
    print(f"F1-score:  {f1_score:.4f} ({f1_score*100:.2f}%)")
    print(f"Accuracy:  {accuracy:.4f} ({accuracy*100:.2f}%)")

    # Simulate imbalance
    if imbalance_ratio:
        neg_ratio, pos_ratio = imbalance_ratio
        new_total = total_samples
        new_negatives = int(new_total * neg_ratio)
        new_positives = int(new_total * pos_ratio)

        # Worst-case: model predicts all negatives
        worst_case_accuracy = new_negatives / new_total

        print("\n=== Imbalance Simulation ===")
        print(f"Negatives: {new_negatives} ({neg_ratio*100:.1f}%), Positives: {new_positives} ({pos_ratio*100:.1f}%)")
        print(f"Worst-case Accuracy (predict all negatives): {worst_case_accuracy:.4f} ({worst_case_accuracy*100:.2f}%)")
        print("This shows why accuracy alone can be misleading for imbalanced datasets.")

# Example usage with your balanced data
evaluate_binary_classifier(TP=400, TN=450, FP=50, FN=100, total_samples=1000, imbalance_ratio=(0.8, 0.2))


=== Metrics ===
Precision: 0.8889 (88.89%)
Recall:    0.8000 (80.00%)
F1-score:  0.8421 (84.21%)
Accuracy:  0.8500 (85.00%)

=== Imbalance Simulation ===
Negatives: 800 (80.0%), Positives: 200 (20.0%)
Worst-case Accuracy (predict all negatives): 0.8000 (80.00%)
This shows why accuracy alone can be misleading for imbalanced datasets.
